runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -969,6 +970,235 @@ J9::TransformUtil::canFoldStaticFinalField(TR::Compilation *comp, TR::Node* node
+   }
+
+static TR_YesNoMaybe safeToAddFearPointAt(TR::Optimization* opt, TR::TreeTop* tt)
+   {
+   TR::Compilation* comp = opt->comp();
+   if (opt->trace())
+      {
+      traceMsg(comp, "Checking if it is safe to add fear point at n%dn\n", tt->getNode()->getGlobalIndex());
+      }
+
+   int32_t callerIndex = tt->getNode()->getByteCodeInfo().getCallerIndex();
+   if (!cannotAttemptOSRDuring(comp, callerIndex) && !comp->isOSRProhibitedOverRangeOfTrees())
+      {
+      if (opt->trace())
+         {
+         traceMsg(comp, "Safe to add fear point because there is no OSR prohibition\n");
+         }
+      return TR_yes;
+      }
+
+   // Look for an OSR point dominating tt in block
+   TR::Block* block = tt->getEnclosingBlock();
+   TR::TreeTop* firstTT = block->getEntry();
+   while (tt != firstTT)
+      {
+      if (comp->isPotentialOSRPoint(tt->getNode()))
+         {
+         TR_YesNoMaybe result = comp->isPotentialOSRPointWithSupport(tt) ? TR_yes : TR_no;
+         if (opt->trace())
+            {
+            traceMsg(comp, "Found %s potential OSR point n%dn, %s to add fear point\n",
+                     result == TR_yes ? "supported" : "unsupported",
+                     tt->getNode()->getGlobalIndex(),
+                     result == TR_yes ? "Safe" : "Not safe");
+            }
+
+         return result;
+         }
+      tt = tt->getPrevTreeTop();
+      }
+
+   TR_HCRGuardAnalysis* guardAnalysis = runHCRGuardAnalysisIfPossible();
+   if (guardAnalysis)
+      {
+      TR_YesNoMaybe result = guardAnalysis->_blockAnalysisInfo[block->getNumber()]->isEmpty() ? TR_yes : TR_no;
+      if (opt->trace())
+         {
+         traceMsg(comp, "%s to add fear point based on HCRGuardAnalysis\n", result == TR_yes ? "Safe" : "Not safe");
+         }
+
+      return result;
+      }
+
+   if (opt->trace())
+      {
+      traceMsg(comp, "Cannot determine if it is safe\n");
+      }
+   return TR_maybe;
+   }
+
+static bool isVarHandleFolding(TR::Compilation* comp, TR_OpaqueClassBlock* declaringClass, char* fieldSignature, int32_t fieldSigLength)
+   {
+   if (comp->getMethodSymbol()->hasMethodHandleInvokes()
+       && !TR::Compiler->cls.classHasIllegalStaticFinalFieldModification(declaringClass))
+      {
+      if (fieldSigLength == 28 && !strncmp(fieldSignature, "Ljava/lang/invoke/VarHandle;", 28))
+         return true;
+      }
+
+   return false;
+   }
+
+/** \brief
+ *     Try to fold static final field with protection
+ *
+ *  \param opt
+ *     The current optimization object.
+ *
+ *  \param currentTree
+ *     The tree with the load of static final field.
+ *
+ *  \param node
+ *     The node which is a load of a static final field.
+ *
+ *  \param varHandleOnly
+ *     True if only folding varHandle static final fields.
+ *     Faslse if folding all static final fileds.
+*/
+bool J9::TransformUtil::attemptStaticFinalFieldFolding(TR::Optimization* opt, TR::TreeTop * currentTree, TR::Node *node, bool varHandleOnly)
```

Comment:
I am not a fan of API designs that need that bool to be passed in all over the optimizer - the bools are hard to interpret and a generally bad design. It would be better to create two public API methods one called attemptVarHandleStaticFinalFieldFolding and attemptGeneralStaticFinalFieldFolding or something like that and have them call the internal implementation with the true or false. That would make it much easier to see in the code what kind of folding is being requested without having to know the signature of the method and what bools are in what order. I could see more bools possibly being added in the future which is why I don't really like this kind of design.

Reply:
Added new API `attemptGenericStaticFinalFieldFolding` and `attemptVarHandleStaticFinalFieldFolding`